# Adaptive gPCA

## Description

Adaptive gPCA, described in
[Fukuyama 2017](https://arxiv.org/abs/1702.00501), is a flexible
method for incorporating side information into principal components
analysis. It was developed for using information about the
phylogenetic structure of bacteria in microbiome data analysis, but it
is applicable to more general kinds of structure. 

## Installation

Adaptive gPCA is implemented in the R package adaptiveGPCA, which can
be installed either from CRAN or github. To install from CRAN
(recommended), use
```r
install.packages("adaptiveGPCA")
```

To install from GitHub, first install devtools, and then use
```r
devtools::install_github("jfukuyama/adaptiveGPCA")
```

## Quick start for microbiome data

The package comes with a [full vignette](https://cran.r-project.org/web/packages/adaptiveGPCA/vignettes/adaptive_gpca_vignette.html), which explains all of the functions and their arguments. To get a feel for what the package does, you can try out the following commands, which run adaptive gPCA on an example microbiome dataset that comes with the package.
```{r}
## Load the required packages
library(adaptiveGPCA)
library(ggplot2)
library(phyloseq)
```
```{r}
## Load the example data, which is stored as a phyloseq object
data(AntibioticPhyloseq)
## Process the phyloseq object so it contains the elements required for adaptive gPCA
pp = processPhyloseq(AntibioticPhyloseq)
## The next line will take a while to run (around a minute on a macbook pro)
out.ff = gpcaFullFamily(pp$X, pp$Q, k = 2)
```
```{r, eval = FALSE}
## This line will open a browser window where you can interactively
## choose the structure parameter for the phylogeny. 	
out.agpca = visualizeFullFamily(out.ff,
                    sample_data = sample_data(AntibioticPhyloseq),
                    sample_mapping = aes(x = Axis1, y = Axis2, color = type),
                    var_data = tax_table(AntibioticPhyloseq),
                    var_mapping = aes(x = Axis1, y = Axis2, color = Phylum))
```
```{r fig.width=6, fig.height=4}
## Runs adaptive gPCA, autamatic choice of structure parameter
out.agpca = adaptivegpca(pp$X, pp$Q, k = 2)
## Plot the samples from adaptive gPCA
ggplot(data.frame(out.agpca$U, sample_data(AntibioticPhyloseq))) +
    geom_point(aes(x = Axis1, y = Axis2, color = type, shape = ind))
```{r fig.width = 8, fig.height = 4}
## Plot the variables from adaptive gPCA
ggplot(data.frame(out.agpca$QV, tax_table(AntibioticPhyloseq))) +
    geom_point(aes(x = Axis1, y = Axis2, color = Phylum))
```
```{r, eval = FALSE}
## This line will open a browser window that allows you to
## access more information about the taxa in the biplot
t = inspectTaxonomy(out.agpca, AntibioticPhyloseq)
```
